codalab/lib/bundle_cli.py

```diff
@@ -815,8 +815,15 @@ def collapse_bare_command(argv):
         """
         try:
             i = argv.index('---')
-            # Convert the command after '---' to a shell-escaped version of the string.
-            shell_escaped_command = [quote(x) for x in argv[i + 1 :]]
+            # This is to get the default list of punctuation chars from shlex. When punctuation_chars = True,
+            # the class property punctuation_chars is set as '();<>|&'. Otherwise, an empty string will be returned.
+            punctuation_chars = list(
+                shlex.shlex(COMMAND_PLACEHOLDER, punctuation_chars=True).punctuation_chars
+            )
+            # Convert string that is not a punctuation char after '---' to a shell-escaped version of the string.
+            shell_escaped_command = [
+                shlex.quote(x) if x not in punctuation_chars else x for x in argv[i + 1 :]
```

Comment:
why are we ignoring only punctuation characters by themselves?  Couldn't they show up as part of an argument (e.g., "hello;")?  I think I'm missing what problem we're trying to solve...  I was thinking that every `x` should just be quoted.

Reply:
For example, we are trying to run `cl run --- date; sleep 10`, in `sys.argv`, the parameters will be changed to `argv = ['run', 'date', ';', 'sleep', '10']`. If we don't leave the punctuation characters out and `shlex.quote` on every single subcommand, we will get the new `argv` as follows:
```
>>> argv = ['run', 'date', ';', 'sleep', '10']
>>> shell_excaped_argv = [shlex.quote(x) for x in argv]
>>> shell_excaped_argv
['run', 'date', "';'", 'sleep', '10']
```
Note that there is an additional pair of quote wrapped around semicolon. Then when it gets executed on real system, we get the following result:
<img width="789" alt="Screen Shot 2020-05-12 at 7 17 18 PM" src="https://user-images.githubusercontent.com/1483372/81764247-fb8b6800-9485-11ea-8298-55b781cd54e7.png">

Basically, shell won't be able to recognize semicolon as a command separator if it wrapped with single quotes.

Reply:
I think if you're using `---`, you should expect each subsequent argument to be treated as actual arguments in a command.  If you want a shell command (e.g., with `;`, `&`, etc.), you should just quote the entire command and not use `---`.  Otherwise I think the logic gets too complicated.  Is there a reason that this wouldn't work?

Reply:
I think it's definitely better if users can properly quote their commands. This change primarily come up due to the issue @kerem discovered when he was using `cli/command` endpoint to submit shell command. And if a user doesn't quote his/her command properly, whitespace splitting may end up grouping argument in the wrong way. 
If we want to keep the way as it is, maybe we could add some comments about properly quoting all the arguments?

Reply:
What's the exact use case that triggered a problem?  Is this from the web terminal?

Reply:
If I am not wrong, he submitted his bundles using a rest api call in his code. @bkgoksel please correct me if I am wrong.

Reply:
@percyliang I think the real issue here that the command parsing rules from rest API (`cli/command`) endpoint is different from CLI. For instance,
1. In CLI, `cl run --- date; sleep 10` will be interpreted as `cl run --- date`. `sleep 10` will be ignored after the command separator (semicolon).
1. In rest API `cli/command` endpoint, when I submit a bundle from the web interface using the following patterns: 
    1. `cl run --- date; sleep 10` will be interpreted as `'date;' sleep 10`.
    1. `cl run --- date;sleep 10` (no whitespace after semicolon) will be interpreted as `'date;sleep' 10`
    1. `cl run --- 'date; sleep 10'` will be interpreted as `'date; sleep 10'`
    All of the above cases can't be executed correctly.